Create Repeat record for each data source row change

SmittieC opened the pull request:
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-3066)

### Note
This is essentially the same changes that [this PR](https://github.com/dimagi/commcare-hq/pull/33781) has. That PR became bloated, so I decided to redo the commits.

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
- Added a new Repeater Type - `DataSourceRepeater`. This repeater is used to forward any data source changes to a client.
- Added a new data source changed signal that gets called when upserting or deleting something from a datasource table.
- When a data source is changed, we check if the data source is "subscibed" to by checking if a Repeater with this data source id exists. If `True` (and the `SUPERSET_ANALYTICS` flag is enabled), we create a RepeatRecord for each row change.

### Some things to note
1. Since we store the `doc_id` in the `payload_id` on a `RepeatRecord`, it points to the row(s) in the datasource (specified on its `Repeater`). This means that when the `RepeatRecord` fires, only then will it go look up the value of that row, meaning that we will always send the latest data / value.
2. Every change creates a repeat record for that `doc_id`, meaning that multiple records can be created that points to the same doc id. While this is not ideal, is not an issue, since we'll effectively just forward the same data to the client multiple times. This can be addressed later on.
3. Currently RepeatRecords are not being cleaned up, meaning we will eventually have **a lot** of RepeatRecords. This will be addressed after the migration to SQL.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
[SUPERSET_ANALYTICS](https://www.commcarehq.org/hq/flags/edit/superset-analytics/)

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
- Local testing
- Unit testing
- Staging testing confirms that the repeater sends the payload in the correct format.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
This cannot be QA'd in a meaningful manner yet.

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations

We might want to remove
- the new `DataSourceRepeater` table

Not doing that should not break anything though

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Charl1996 committed:
Create connection settings when service subscribes to data source changes

Charl1996 committed:
Put API behind FF

Charl1996 committed:
Add repeater and change connection settings name

Charl1996 committed:
Add tests

Charl1996 committed:
Don't require refresh token on subscribe-to-datasource endpoint

Charl1996 committed:
Consider token expiry for requesting client_credentials oauth token

Charl1996 committed:
Small refactor: use update_or_create

Charl1996 committed:
Remove user on tearDownClass

Charl1996 committed:
Move imports to top

Charl1996 committed:
Revert "Consider token expiry for requesting client_credentials oauth token"

This reverts commit 42284274afd753184bb8388c61ec820dfb146b2a.

SmittieC committed:
Small Repeater refactor

SmittieC committed:
Add DataSourceRepeater

SmittieC committed:
Add DataSourceRepeater test

SmittieC committed:
Add method to IndicatorSqlAdapter to fetch docs from a datasource

SmittieC committed:
Add method that calls a signal to create data source change repeat records

SmittieC committed:
Formatting

SmittieC committed:
Add ucr_data_source_updated signal and call it when datasource changes

SmittieC committed:
Update pillow tests to test the signal calls

SmittieC committed:
Update function name

snopoke commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -216,6 +228,30 @@ def doc_exists(self, doc):
             query = session.query(self.get_table()).filter_by(doc_id=doc['_id'])
             return session.query(query.exists()).scalar()
 
+    def get_docs(self, doc_id):
```
nit: rename
```suggestion
    def get_rows(self, doc_id):
```

snopoke commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -216,6 +228,30 @@ def doc_exists(self, doc):
+            for row in query:
+                yield row
+
+        table_ = get_table(query)
+        headers = next(table_)
+
+        named_entries = []
+        for row in table_:
+            columns_data = {}
+            for column_name, column_value in zip(headers, row):
+                columns_data[column_name] = column_value
+            named_entries.append(columns_data)
```
shouldn't this be within the session context?

SmittieC replied:
Interesting. Yes, but no? Somehow it worked...I'll move it into the session though..

snopoke commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -216,6 +228,30 @@ def doc_exists(self, doc):
+                yield row
+
+        table_ = get_table(query)
+        headers = next(table_)
```
you could just remove the first yield from `get_table` and use it directly here:
```python
headers = list(table.columns.keys())
``` 

I think that makes `get_table` redundant

SmittieC replied:
https://github.com/dimagi/commcare-hq/pull/33877/commits/e0c5698e49eb7d6fc68053757a533582e245926d

snopoke commented on corehq/apps/userreports/util.py:
```diff
@@ -355,3 +369,21 @@ def unescape(value):
+            return
+
+        for row_change in row_changes:
+            update_log = DataSourceUpdateLog(domain, data_source_id=data_source_id, doc_id=row_change["doc_id"])
```
Since you only need doc IDs you could just pass a list of doc IDs instead of `row_changes` dict.

Also you should remove duplicates (for cases where a single doc creates multiple rows).

snopoke commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -177,6 +177,12 @@ def save_rows(self, rows, use_shard_col=True):
+        register_data_source_row_change(
+            domain=self.config.domain,
+            data_source_id=self.config._id,
+            row_changes=[{key: str(value) for key, value in row.items()} for row in formatted_rows],
```
I think this could just use `doc_ids` variable defined above

SmittieC replied:
Ha-ha, good catch!

SmittieC replied:
https://github.com/dimagi/commcare-hq/pull/33877/commits/395a4627d252ca929a3233135753333ea55b8771

snopoke commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -208,6 +214,12 @@ def bulk_delete(self, docs, use_shard_col=True):
+            row_changes=[{"doc_id": doc['_id']} for doc in docs],
+        )
+
     def delete(self, doc, use_shard_col=True):
```
~~It looks like single doc deletes don't use this:~~

~~https://github.com/dimagi/commcare-hq/blob/6b0021c4a3b753a8beeac2f0e748abf2c575b6b7/corehq/apps/userreports/pillow.py#L540~~

Scratch that, it's just bad variable naming.

snopoke commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -208,6 +214,12 @@ def bulk_delete(self, docs, use_shard_col=True):
+        register_data_source_row_change(
+            domain=self.config.domain,
+            data_source_id=self.config._id,
+            row_changes=[{"doc_id": doc['_id']} for doc in docs],
```
could just use `doc_ids` var already defined

SmittieC replied:
https://github.com/dimagi/commcare-hq/pull/33877/commits/6c56887c3067614a0eeff90561e85e234ec657b8

snopoke reviewed (commented):
Generally looks good. Few minor comments.

SmittieC committed:
Pass doc_ids to register_data_source_row_change directly

SmittieC committed:
Change method name from get_docs to get_rows

SmittieC committed:
Remove duplicate doc_ids

SmittieC committed:
Refactor get_rows method

SmittieC committed:
Small refactor

Charl1996 reviewed (approved):
Good from my side.

mkangia commented on corehq/motech/repeaters/repeater_generators.py:
```diff
@@ -739,3 +740,20 @@ def content_type(self):
+    def content_type(self):
+        return 'application/json'
+
+    def get_payload(self, repeat_record, docs: list):
```
isn't this just a doc and not docs?

kaapstorm replied:
In `get_payload()` and `payload_doc()` the payload is a list of dictionaries, but in `allowed_to_forward()` the payload is a `DataSourceUpdateLog` instance. This felt surprising, and inconsistent with other Repeater subclasses. So I changed them to all use a `DataSourceUpdateLog` instance. ff853d68d601ee0e8bb65ff365e53b3dd935a5be

SmittieC replied:
Nice!

mkangia commented on corehq/motech/repeaters/tests/test_repeater.py:
```diff
@@ -1284,6 +1287,65 @@ def test_five_retries(self):
             self.assertEqual(interval, timedelta(hours=expected_interval_hours))
 
 
+class DataSourceRepeaterTest(BaseRepeaterTest, TestXmlMixin):
```
why this inheriting `TextXmlMixin`?

kaapstorm replied:
2c7d316dd79acca4967526f08bedd20ca0887451

mkangia commented on corehq/motech/repeaters/tests/test_repeater.py:
```diff
@@ -1284,6 +1287,65 @@ def test_five_retries(self):
+        payload = self.repeater.get_payload(repeat_record)
+        payload = json.loads(payload)
+        self.assertEqual(payload["data_source_id"], self.data_source_id)
+        self.assertTrue(isinstance(payload["data"], list))
```
can we test for the actual data in it?

kaapstorm replied:
2c7d316dd79acca4967526f08bedd20ca0887451

mkangia commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -216,6 +228,24 @@ def doc_exists(self, doc):
+        table = self.get_table()
+        headers = list(table.columns.keys())
+
+        named_entries = []
```
nit: this naming seems confusing. I believe you could simply call this rows.

kaapstorm replied:
c43a584ad43bf8dddffed31948287f2fecd0bf98

mkangia commented on corehq/apps/userreports/sql/adapter.py:
```diff
@@ -216,6 +228,24 @@ def doc_exists(self, doc):
+
+        named_entries = []
+        with self.session_context() as session:
+            query = session.query(self.get_table()).filter_by(doc_id=doc_id)
```
nit: ideally there should only be one row right? should we raise an error if there is more than 1?

kaapstorm replied:
There are cases where a document can result in multiple rows.

mkangia replied:
Can you say more?
I would think one form/case would have only one row in the table.

kaapstorm replied:
I thought so too, but [there are ways to iterate form/case data](https://commcare-hq.readthedocs.io/ucr.html#saving-multiple-rows-per-case-form).

kaapstorm replied:
We've confirmed that since we only fetch the rows when the repeat record is sent, and the payload structure supports multiple rows for a doc_id, there is no problem with this approach.

mkangia commented on corehq/motech/repeaters/models.py:
```diff
@@ -852,6 +852,41 @@ def get_all_repeater_types():
+    payload_generator_classes = (DataSourcePayloadGenerator,)
+
+    def allowed_to_forward(self, update_log):
+        return update_log.data_source_id == self.data_source_id
```
I see now that it is `DataSourceUpdateLog`. Why did we need to add this log dataclass? because we can pass only one object but there was a need to pass around data source ID and payload id as well?

kaapstorm replied:
Yeah, that's exactly it. `allowed_to_forward()` is passed a "payload" (usually a form, case, etc.) and has to determine from it whether this repeater is responsible for it and should forward it. We won't be able to figure that out from just a form submission ID or a case ID, so `DataSourceUpdateLog` serves as the payload.

mkangia commented on corehq/motech/repeaters/models.py:
```diff
@@ -847,6 +852,41 @@ def get_all_repeater_types():
+        datasource_adapter = get_indicator_adapter(config, load_source='repeat_record')
+        return datasource_adapter.get_rows(repeat_record.payload_id)
+
+    def clear_caches(self):
```
Should this be called when a repeater is deleted as well?

kaapstorm replied:
746661b8d502eecde09180224bc012347b5a448d

mkangia reviewed (approved):
This is great. Thanks for adapting to the new approach so quickly @SmittieC 

Non blocking comments that can be taken in a follow up once you are back.
Going to hold off for a day or two to see if @kaapstorm wants to take a look as well and then merge it.

kaapstorm committed:
Rename, refactor IndicatorSqlAdapter.get_rows()

kaapstorm committed:
Clear caches on delete

kaapstorm committed:
isort

kaapstorm committed:
DataSourceRepeater payload is DataSourceUpdateLog

Use the same type for payload every where (instead of
DataSourceUpdateLog in one place, and a list of dictionaries in a
different place). Requires adding `rows` attribute to
DataSourceUpdateLog.

kaapstorm committed:
Update test: Assert whole payload

kaapstorm commented:
(The 10 commits by Charl1996 come from PR https://github.com/dimagi/commcare-hq/pull/33795 which was merged into the old, dropped, branch.)

mkangia commented on corehq/motech/repeaters/models.py:
```diff
@@ -201,6 +201,10 @@ def save(self, *args, **kwargs):
         return super().save(*args, **kwargs)
 
+    def delete(self, *args, **kwargs):
+        self.clear_caches()
```
:+1:

mkangia commented on corehq/motech/repeaters/models.py:
```diff
@@ -861,6 +861,12 @@ def get_all_repeater_types():
 class DataSourceRepeater(Repeater):
+    """
+    Forwards the UCR data source rows that are updated by a form
+    submission or a case update.
```
nit: creation/update?

kaapstorm replied:
3aca2661ebc995c3aa5c91b31be992e01b8d031d

mkangia commented on corehq/apps/userreports/views.py:
```diff
@@ -1536,6 +1538,32 @@ def get_table(q):
+@api_auth()
+@require_permission(HqPermissions.view_reports)
+@api_throttle
+def subscribe_to_data_source_changes(request, config_id):
```
what is this change?

mkangia replied:
Should we put this in a separate PR? It seems out of scope of what this PR was addressing and seems to have needed quite a few commits.

kaapstorm replied:
Sorry for the confusion. Yes, that chunk of work was in [a separate PR](https://github.com/dimagi/commcare-hq/pull/33795), which has been merged into this branch.

mkangia commented:
> (The 10 commits by Charl1996 come from PR https://github.com/dimagi/commcare-hq/pull/33795 which was merged into the old, dropped, branch.)

Oh no, what just happened here?

kaapstorm committed:
Don't reate-limit tests

kaapstorm committed:
Small clarification in docstring

mkangia commented:
Thanks for the followup @kaapstorm 
Please share your thoughts on https://github.com/dimagi/commcare-hq/pull/33877#discussion_r1432512098
Seems blocking.

zandre-eng reviewed (approved):
Looks good.
